ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/forkchoice/ForkChoiceNotifier.java

```diff
@@ -0,0 +1,289 @@
+        .orElse(false);
+  }
+
+  public void onForkChoiceUpdated(
+      Bytes32 optimisticHeadBeaconBlockRoot, ForkChoiceState forkChoiceState) {
+    LOG.debug("onForkChoiceUpdated {}, {}", optimisticHeadBeaconBlockRoot, forkChoiceState);
+
+    Optional<Bytes32> currentValidatedHead =
+        recentChainData.getChainHead().map(StateAndBlockSummary::getRoot);
+    boolean isForkChoiceStateOnValidatedHead =
+        currentValidatedHead.isPresent()
+            && optimisticHeadBeaconBlockRoot.equals(currentValidatedHead.get());
+
+    currentForkChoiceUpdatedInfo.forkChoiceState = Optional.of(forkChoiceState);
+    currentForkChoiceUpdatedInfo.isOnValidatedHead = isForkChoiceStateOnValidatedHead;
+
+    recentChainData
+        .getCurrentSlot()
+        .ifPresent(
+            (slot) -> {
+              UInt64 targetSlot = slot.plus(1);
+              callExecutionEngineForkChoiceUpdated(getPayloadAttributes(targetSlot), targetSlot);
+            });
+  }
+
+  public void onAttestationsDue(UInt64 slot) {
+    LOG.debug("onAttestationsDue {}", slot);
+
+    maybeLastAttestationDueSlot = Optional.of(slot);
+
+    UInt64 targetSlot = slot.plus(1);
+
+    if (currentForkChoiceUpdatedInfo.hasNotBeenCalledForSlot(targetSlot)) {
+
+      Optional<PayloadAttributes> payloadAttributes = getPayloadAttributes(targetSlot);
+
+      if (payloadAttributes.isPresent()) {
+        callExecutionEngineForkChoiceUpdated(payloadAttributes, targetSlot);
+      }
+    }
+  }
+
+  public CurrentForkChoiceUpdatedInfo getCurrentForkChoiceUpdatedInfo() {
+    return currentForkChoiceUpdatedInfo;
+  }
+
+  private void updatePreparableProposers(
+      Collection<BeaconPreparableProposer> beaconPreparableProposers, UInt64 currentSlot) {
+
+    final int maxProposerSeenSlot = spec.getSlotsPerEpoch(currentSlot) * MAX_PROPOSER_SEEN_EPOCHS;
+    boolean newProposerIndexAdded = false;
+
+    // remove expired proposers
+    proposerIndexLastSeenSlotAndFeeRecipient =
+        proposerIndexLastSeenSlotAndFeeRecipient.entrySet().stream()
+            .filter(p -> p.getValue().lastSeenSlot.compareTo(currentSlot) < maxProposerSeenSlot)
+            .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+
+    // update or add new proposers
+    for (BeaconPreparableProposer beaconPreparableProposer : beaconPreparableProposers) {
+      ProposerLastSeenSlotAndFeeRecipient current =
+          proposerIndexLastSeenSlotAndFeeRecipient.get(
+              beaconPreparableProposer.getValidatorIndex());
+      if (current == null) {
+        proposerIndexLastSeenSlotAndFeeRecipient.put(
+            beaconPreparableProposer.getValidatorIndex(),
+            new ProposerLastSeenSlotAndFeeRecipient(
+                currentSlot, beaconPreparableProposer.getFeeRecipient()));
+
+        newProposerIndexAdded = true;
+
+      } else {
+        current.lastSeenSlot = currentSlot;
+      }
+    }
```

Comment:
We need to update the feeRecipient to the latest value as well.  I'd be tempted to just do:
```suggestion
    for (BeaconPreparableProposer beaconPreparableProposer : beaconPreparableProposers) {
      proposerIndexLastSeenSlotAndFeeReceipient.put(
          beaconPreparableProposer.getValidatorIndex(),
          new ProposerLastSeenSlotAndFeeRecipient(currentSlot, beaconPreparableProposer.getFeeRecipient());
    }
```

Reply:
oh good catch here.